include/zephyr/lorawan/lorawan.h

```diff
@@ -330,6 +338,29 @@ void lorawan_get_payload_sizes(uint8_t *max_next_payload_size,
  */
 int lorawan_set_region(enum lorawan_region region);
 
+/**
+ * @brief Request for time according to DeviceTimeReq MAC cmd
+ *
+ * Append MAC DevTimeReq command. It will be processed on next send
+ * message or force sending empty message to request time immediately.
+ *
+ * @param force_request Immediately send an empty message to execute the request
+ * @return 0 if successful, negative errno otherwise
+ */
+int lorawan_request_device_time(bool force_request);
+
+/**
+ * @brief Retrieve the current time from LoRaWAN stack updated by
+ * DeviceTimeAns on MAC layer.
+ *
+ * The epoch starts at 1970-01-01T00:00:00Z.
+ *
+ * @param epoch_time Synchronized time in epoch format truncated to 32-bit.
+ *
+ * @return 0 if successful, -EAGAIN if the time is not yet updated.
+ */
+int lorawan_device_time_get(uint32_t *epoch_time);
```

Comment:
I suggest to mention clearly that this is UNIX epoch and not GPS epoch (which is used in AppClockSync):

```suggestion
/**
 * @brief Retrieve the current UNIX timestamp from LoRaWAN stack updated by
 * DeviceTimeAns on MAC layer.
 *
 * The UNIX epoch starts at 1970-01-01T00:00:00Z.
 *
 * @param unix_time Synchronized time in seconds truncated to 32-bit.
 *
 * @return 0 if successful, -EAGAIN if the time is not yet updated.
 */
int lorawan_device_time_get(uint32_t *unix_time);
```

Reply:
Are you sure about that? As shown in an earlier screenshot, section 5.9 of the spec says:

> (*) The GPS epoch (i.e., January 6, 1980 00:00:00 UTC) is used as origin. The seconds field is the number of seconds elapsed since the origin. This field increases monotonically by 1 every second. To convert this field to UTC time, the leap seconds SHALL be taken into account

Reply:
The timestamp is transferred in GPS epoch format, but this function gets the time through [SysTimeGet](https://github.com/zephyrproject-rtos/loramac-node/blob/842413c5fb98707eb5f26e619e8e792453877897/src/system/systime.h#L122), which uses the UNIX epoch.

See here for the code section doing the conversion: https://github.com/zephyrproject-rtos/loramac-node/blob/842413c5fb98707eb5f26e619e8e792453877897/src/mac/LoRaMac.c#L2552-L2578

However, it looks like it is ignoring the leap seconds...

Reply:
Great, so the time is only out by 20 something seconds. I would advocate for translating the value back to GPS time in our function then. Let the user perform conversions starting from a correct time, together with any knowledge of leap seconds they may have.

Reply:
> Now that Zephyr has a RTC API we could use both DeviceTimeReq and AppClockSync to set the time of the RTC instead of coming up with custom clock/time APIs based on uncommon epochs. Thoughts?

Not really a fan, are we adding `gmtime` into our libraries now? It makes more sense to me to leave it to the user to perform some action. Even using the RTC API would require providing the device, so it couldn't be automatic either.